Tools/SandboxTest.ps1

```diff
@@ -0,0 +1,298 @@
+}
+
+function Get-LatestHash {
+  $shaUrl = ((Invoke-WebRequest $apiLatestUrl -UseBasicParsing | ConvertFrom-Json).assets | Where-Object { $_.name -match '^Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.txt$' }).browser_download_url
+
+  $shaFile = Join-Path -Path $tempFolder -ChildPath 'Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.txt'
+  $WebClient.DownloadFile($shaUrl, $shaFile)
+
+  Get-Content $shaFile
+}
+
+# Hide the progress bar of Invoke-WebRequest
+$oldProgressPreference = $ProgressPreference
+$ProgressPreference = 'SilentlyContinue'
+
+$desktopAppInstaller = @{
+  fileName = 'Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.appxbundle'
+  url      = $(Get-LatestUrl)
+  hash     = $(Get-LatestHash)
+}
+
+$ProgressPreference = $oldProgressPreference
+
+$vcLibsUwp = @{
+  fileName = 'Microsoft.VCLibs.x64.14.00.Desktop.appx'
+  url      = 'https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx'
+  hash     = '6602159c341bafea747d0edf15669ac72df8817299fbfaa90469909e06794256'
+}
+
+$dependencies = @($desktopAppInstaller, $vcLibsUwp)
+
+# Clean temp directory
+
+Get-ChildItem $tempFolder -Recurse -Exclude $dependencies.fileName | Remove-Item -Force -Recurse
+
+if (-Not [String]::IsNullOrWhiteSpace($Manifest)) {
+  Copy-Item -Path $Manifest -Recurse -Destination $tempFolder
+}
+
+# Download dependencies
+
+Write-Host '--> Checking dependencies'
+
+$desktopInSandbox = 'C:\Users\WDAGUtilityAccount\Desktop'
+
+foreach ($dependency in $dependencies) {
+  $dependency.file = Join-Path -Path $tempFolder -ChildPath $dependency.fileName
+  $dependency.pathInSandbox = Join-Path -Path $desktopInSandbox -ChildPath (Join-Path -Path $tempFolderName -ChildPath $dependency.fileName)
+
+  # Only download if the file does not exist, or its hash does not match.
+  if (-Not ((Test-Path -Path $dependency.file -PathType Leaf) -And $dependency.hash -eq $(get-filehash $dependency.file).Hash)) {
+    Write-Host @"
+    - Downloading:
+      $($dependency.url)
+"@
+
+    try {
+      $WebClient.DownloadFile($dependency.url, $dependency.file)
+    }
+    catch {
+      throw "Error downloading $($dependency.url)."
+    }
+    if (-not ($dependency.hash -eq $(get-filehash $dependency.file).Hash)) {
+      throw 'Hashes do not match, try gain.'
+    }
+  }
+}
+
+Write-Host
+
+# Create Bootstrap script
+
+# See: https://stackoverflow.com/a/14382047/12156188
+$bootstrapPs1Content = @'
+function Update-Environment {
+  $locations = 'HKLM:\SYSTEM\CurrentControlSet\Control\Session Manager\Environment',
+               'HKCU:\Environment'
+
+  $locations | ForEach-Object {
+    $k = Get-Item $_
+    $k.GetValueNames() | ForEach-Object {
+      $name  = $_
+      $value = $k.GetValue($_)
+
+      if ($userLocation -and $name -ieq 'PATH') {
+        $Env:Path += ";$value"
+      } else {
+        Set-Item -Path Env:$name -Value $value
+      }
+    }
+
+    $userLocation = $true
+  }
+}
+
+
+'@
+
+$bootstrapPs1Content += @"
+Write-Host @'
+--> Installing WinGet
+
+
+
+
+

```

Comment:
Why so much whitespace?

Reply:
Because of this (the progress bar stays on the screen even after the completion):

![U6ho8iV7OL](https://user-images.githubusercontent.com/29582865/90988322-56922800-e568-11ea-81f9-a32f191c35f9.gif)

But this is not happening anymore in my tests on 21H1. At the time the GIF were created I think I tested with 2004, so I don't know how it is for 20H2 (the current stable Windows version).

One option would be to hide the progress bar though, but I just wanted to make the script as transparent as possible. I even though about turning on the PowerShell debug feature to output each command being runned, but it turned out to be too noisy.

What do you think?

Reply:
Because of this (the progress bar stays on the screen even after the completion):

![U6ho8iV7OL](https://user-images.githubusercontent.com/29582865/90988322-56922800-e568-11ea-81f9-a32f191c35f9.gif)

But this is not happening anymore in my tests on 21H1. At the time the GIF were created I think I tested with 2004, so I don't know how it is for 20H2 (the current stable Windows version).

One option would be to hide the progress bar though, but I just wanted to make the script as transparent as possible. I even though about turning on the PowerShell debug feature to output each command being runned, but it turned out to be too noisy.

What do you think?

Reply:
Gotcha, keep it then no prob